File: market-application/src/main/java/com/example/market/application/service/OrderBookQueryService.java
package com.example.market.application.service;

import com.example.market.application.port.in.OrderBookQueryUseCase;
import com.example.market.application.port.out.OrderBookQueryPort;
import com.example.market.domain.catalog.SkuId;
import com.example.market.domain.shared.Money;
import com.example.market.domain.trading.Bid;
import com.example.market.domain.trading.Listing;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Clock;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;

/**
 * 호가창 view — read only. 같은 가격대의 호가를 PriceLevel 로 집계.
 */
@Service
@RequiredArgsConstructor
public class OrderBookQueryService implements OrderBookQueryUseCase {

    private final OrderBookQueryPort orderBook;
    private final Clock clock;

    @Override
    @Transactional(readOnly = true)
    public OrderBookView view(SkuId skuId, int depth) {
        var now = clock.instant();
        List<Listing> asks = orderBook.topNAsks(skuId, depth, now);
        List<Bid> bids = orderBook.topNBids(skuId, depth, now);

        Optional<Money> lowestAsk = asks.stream().findFirst().map(Listing::askPrice);
        Optional<Money> highestBid = bids.stream().findFirst().map(Bid::bidPrice);

        return new OrderBookView(skuId, lowestAsk, highestBid,
                aggregateAsks(asks), aggregateBids(bids));
    }

    private List<PriceLevel> aggregateAsks(List<Listing> asks) {
        Map<Money, Integer> counts = new TreeMap<>();      // ASK 는 가격 ↑
        for (Listing a : asks) counts.merge(a.askPrice(), 1, Integer::sum);
        return counts.entrySet().stream()
                .map(e -> new PriceLevel(e.getKey(), e.getValue()))
                .toList();
    }

    private List<PriceLevel> aggregateBids(List<Bid> bids) {
        Map<Money, Integer> counts = new TreeMap<>(java.util.Comparator.reverseOrder()); // BID 는 가격 ↓
        for (Bid b : bids) counts.merge(b.bidPrice(), 1, Integer::sum);
        return counts.entrySet().stream()
                .map(e -> new PriceLevel(e.getKey(), e.getValue()))
                .toList();
    }
}

File: market-application/src/main/kotlin/com/example/market/application/service/OrderBookQueryService.kt
package com.example.market.application.service

import com.example.market.application.port.`in`.OrderBookQueryUseCase
import com.example.market.application.port.out.OrderBookQueryPort
import com.example.market.domain.catalog.SkuId
import com.example.market.domain.shared.Money
import com.example.market.domain.trading.Bid
import com.example.market.domain.trading.Listing
import java.time.Clock
import java.util.Optional
import java.util.TreeMap
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * 호가창 view — read only. 같은 가격대의 호가를 PriceLevel 로 집계.
 */
@Service
@Transactional(readOnly = true)
open class OrderBookQueryService(
    private val orderBook: OrderBookQueryPort,
    private val clock: Clock,
) : OrderBookQueryUseCase {

    override fun view(skuId: SkuId, depth: Int): OrderBookQueryUseCase.OrderBookView {
        val now = clock.instant()
        val asks: List<Listing> = orderBook.topNAsks(skuId, depth, now)
        val bids: List<Bid> = orderBook.topNBids(skuId, depth, now)

        val lowestAsk: Optional<Money> = Optional.ofNullable(asks.firstOrNull()?.askPrice)
        val highestBid: Optional<Money> = Optional.ofNullable(bids.firstOrNull()?.bidPrice)

        return OrderBookQueryUseCase.OrderBookView(
            skuId, lowestAsk, highestBid,
            aggregateAsks(asks), aggregateBids(bids),
        )
    }

    private fun aggregateAsks(asks: List<Listing>): List<OrderBookQueryUseCase.PriceLevel> {
        val counts: TreeMap<Money, Int> = TreeMap()      // ASK 는 가격 ↑
        for (a in asks) counts.merge(a.askPrice, 1, Int::plus)
        return counts.entries.map { OrderBookQueryUseCase.PriceLevel(it.key, it.value) }
    }

    private fun aggregateBids(bids: List<Bid>): List<OrderBookQueryUseCase.PriceLevel> {
        val counts: TreeMap<Money, Int> = TreeMap(Comparator.reverseOrder())   // BID 는 가격 ↓
        for (b in bids) counts.merge(b.bidPrice, 1, Int::plus)
        return counts.entries.map { OrderBookQueryUseCase.PriceLevel(it.key, it.value) }
    }
}
